lib/charms/operator_libs_linux/v0/apt.py

```diff
@@ -1314,48 +1422,309 @@ def _parse(line: str, filename: str) -> DebianRepository:
-            )
+    May define multiple DebianRepository objects.
+    """
 
-        self._repository_map["{}-{}-{}".format(repo.repotype, repo.uri, repo.release)] = repo
+    def __init__(self, numbered_lines: List[Tuple[int, str]], filename: str = ""):
+        self._filename = filename
+        self._numbered_lines = numbered_lines
+        if not numbered_lines:
+            self._repos = ()
+            self._gpg_key_filename = ""
+            self._gpg_key_from_stanza = None
+            return
+        options, line_numbers = _deb822_stanza_to_options(numbered_lines)
+        repos, gpg_key_info = _deb822_options_to_repos(
+            options, line_numbers=line_numbers, filename=filename
+        )
+        for repo in repos:
+            repo._deb822_stanza = self  # pyright: ignore[reportPrivateUsage]
+        self._repos = repos
+        self._gpg_key_filename, self._gpg_key_from_stanza = gpg_key_info
 
-    def disable(self, repo: DebianRepository) -> None:
-        """Remove a repository. Disable by default.
+    @property
+    def repos(self) -> Tuple[DebianRepository, ...]:
+        """The repositories defined by this deb822 stanza."""
+        return self._repos
 
-        Args:
-          repo: a `DebianRepository` to disable
+    def get_gpg_key_filename(self) -> str:
+        """Return the path to the GPG key for this stanza.
+
+        Import the key first, if the key itself was provided in the stanza.
+        Return an empty string if no filename or key was provided.
         """
-        searcher = "{} {}{} {}".format(
-            repo.repotype, repo.make_options_string(), repo.uri, repo.release
-        )
+        if self._gpg_key_filename:
+            return self._gpg_key_filename
+        if self._gpg_key_from_stanza is None:
+            return ""
+        # a gpg key was provided in the stanza
+        # and we haven't already imported it
+        self._gpg_key_filename = import_key(self._gpg_key_from_stanza)
+        return self._gpg_key_filename
 
-        for line in fileinput.input(repo.filename, inplace=True):
-            if re.match(r"^{}\s".format(re.escape(searcher)), line):
-                print("# {}".format(line), end="")
-            else:
-                print(line, end="")
 
-        self._repository_map["{}-{}-{}".format(repo.repotype, repo.uri, repo.release)] = repo
+class MissingRequiredKeyError(InvalidSourceError):
+    """Missing a required value in a source file."""
+
+    def __init__(self, message: str = "", *, file: str, line: Optional[int], key: str) -> None:
+        super().__init__(message, file, line, key)
+        self.file = file
+        self.line = line
+        self.key = key
+
+
+class BadValueError(InvalidSourceError):
+    """Bad value for an entry in a source file."""
+
+    def __init__(
+        self,
+        message: str = "",
+        *,
+        file: str,
+        line: Optional[int],
+        key: str,
+        value: str,
+    ) -> None:
```

Comment:
These custom exceptions can probably go back to just being InvalidSourceErrors

Reply:
Seems fine to leave as is.